src/main/java/com/odysseusinc/arachne/portal/service/impl/BaseUserServiceImpl.java

```diff
@@ -322,10 +325,90 @@ public void remove(Long id) throws ValidationException, UserNotFoundException, N
         rawUserRepository.delete(user.getId());
     }
 
+    @Override
+    public List<U> createAll(final @NotNull BulkUsersRegistrationDTO bulkUsersDto)
+            throws PasswordValidationException, ValidationException {
+
+        if (bulkUsersDto.getTenantDtos() == null || bulkUsersDto.getTenantDtos().isEmpty()) {
+            throw new ValidationException("tenants: must be not empty");
+        }
+
+        List<U> users = convertRegistrationDTOs(bulkUsersDto);
+        List<U> createdUsers = userRepository.save(users);
+        if (bulkUsersDto.getEmailConfirmationRequired()) {
+            Optional<CommonUserRegistrationDTO> optionalUser = bulkUsersDto.getUserDtos().stream().findFirst();
+            if (!optionalUser.isPresent()) {
+                throw new ValidationException("user: must be not null");
+            }
+            Optional<UserRegistrant> userRegistrant = userRegistrantService.findByToken(optionalUser.get().getRegistrantToken());
+
+            for (U user : createdUsers) {
+                sendRegistrationEmail(user, userRegistrant, optionalUser.get().getCallbackUrl());
+            }
+        }
+
+        return createdUsers;
+    }
+
+    private List<U> convertRegistrationDTOs(BulkUsersRegistrationDTO bulkUsersDto) throws PasswordValidationException {
+
+        Set<Tenant> tenants = bulkUsersDto.getTenantDtos().stream()
+                .map(tenant -> conversionService.convert(tenant, Tenant.class))
+                .collect(Collectors.toSet());
+
+        List<U> users = bulkUsersDto.getUserDtos().stream()
+                .map(dto -> (U) conversionService.convert(dto, User.class))
+                .collect(Collectors.toList());
+
+        for (U user : users) {
+            user.setTenants(tenants);
+            user.setOrigin(UserOrigin.NATIVE);
+            if (!bulkUsersDto.getEmailConfirmationRequired()) {
+                user.setEmailConfirmed(true);
+            } else {
+                user.setEmailConfirmed(false);
+                user.setRegistrationCode(UUID.randomUUID().toString());
+            }
+
+            updateFields(user);
+        }
+
+        return users;
+    }
+
     @Override
     public U create(final @NotNull U user)
             throws NotUniqueException, NotExistException, PasswordValidationException {
 
+        updateFields(user);
+
+        return userRepository.save(user);
+    }
+
+    @Override
+    public U createWithEmailVerification(final @NotNull U user, String registrantToken, String callbackUrl)
```

Comment:
Why is the method marked as new if it existed already? Check codestyle

Reply:
IntelliJ Idea shifted it though the code style was taken from Tech intro